2021/17.py

```diff
@@ -3,22 +3,31 @@
 import heapq
 import itertools
 from collections import defaultdict, Counter, deque
+from math import ceil, sqrt
+
+#target area: x=20..30, y=-10..-5
+# MinX, MaxX = (20, 30)
+# MinY, MaxY = (-10, -5)
 
 #target area: x=96..125, y=-144..-98
+MinX, MaxX = (96, 125)
+MinY, MaxY = (-144, -98)
 
 # y(t) ~ -t^2 + DY*t
 
 p2 = 0
 ans = 0
-for DX in range(150):
-    for DY in range(-150, 1000):
+
+minDX = int(ceil((sqrt(8*MinX+1)-1)/2))
+for DX in range(minDX, MaxX+1):
+    for DY in range(MinY, 2-MinY):
         ok = False
         max_y = 0
         x = 0
         y = 0
         dx = DX
         dy = DY
-        for t in range(1000):
+        while x <= MaxX and y >= MinY:
```

Comment:
Instead of iterating through an arbitrary number of time steps, you can just iterate until the probe steps beyond the target range.